rkt/gc.go

```diff
@@ -197,29 +198,50 @@ func emptyGarbage() error {
 	return nil
 }
 
-func mountPodStage1(ts *treestore.Store, p *pkgPod.Pod) error {
+// mountPodStage1 tries to remount stage1 image overlay in case
+// it is not anymore available in place (e.g. a reboot happened
+// in-between). If an overlay mount is already there, we assume
+// stage1 image is properly in place and we don't perform any
+// further actions. A boolean is returned to signal whether a
+// a new mount was performed.
+func mountPodStage1(ts *treestore.Store, p *pkgPod.Pod) (bool, error) {
 	if !p.UsesOverlay() {
-		return nil
+		return false, nil
 	}
 
 	s1Id, err := p.GetStage1TreeStoreID()
 	if err != nil {
-		return errwrap.Wrap(errors.New("error getting stage1 treeStoreID"), err)
+		return false, errwrap.Wrap(errors.New("error getting stage1 treeStoreID"), err)
 	}
 	s1rootfs := ts.GetRootFS(s1Id)
-
 	stage1Dir := common.Stage1RootfsPath(p.Path())
+
+	if mounts, err := mountinfo.ParseMounts(0); err == nil {
+		for _, m := range mounts {
+			if m.MountPoint == stage1Dir {
+				// stage1 image overlay already in place
+				return false, nil
+			}
+		}
+	}
+
 	overlayDir := filepath.Join(p.Path(), "overlay")
 	imgDir := filepath.Join(overlayDir, s1Id)
 	upperDir := filepath.Join(imgDir, "upper")
 	workDir := filepath.Join(imgDir, "work")
 
 	opts := fmt.Sprintf("lowerdir=%s,upperdir=%s,workdir=%s", s1rootfs, upperDir, workDir)
-	if err := syscall.Mount("overlay", stage1Dir, "overlay", 0, opts); err != nil {
-		return errwrap.Wrap(errors.New("error mounting stage1"), err)
+	err = syscall.Mount("overlay", stage1Dir, "overlay", 0, opts)
+	if err != nil {
+		// -EBUSY means the overlay mount is already present.
+		// This behavior was introduced in Linux 4.13, double-mounts were allowed
+		// in older kernels.
+		if err == syscall.EBUSY {
+			return false, nil
```

Comment:
Shouldn't the double-mount be detected above by the `if m.MountPoint == stage1Dir`? In this case, the EBUSY could be a real error since the code above would have detected it.

Reply:
True, this is probably a tautological check which won't be reached in normal cases. However I trust the kernel mount return code more than our mountinfo parser, so this helps avoiding GC being aborted by mistake in situations were it could have succeeded.
In the best cases this let the kernel catch corner-cases we didn't see, in the worst case we will error our in the next step trying to call stage1 GC.

Reply:
I think there's a good argument for at least logging it even if we don't fail on it.